Add u: options namespace (#846)
unicode-org/message-format-wg: spec/u-namespace.md

# MessageFormat 2.0 Unicode Namespace

The `u:` namespace is reserved for use by the Unicode Consortium.

## Options

This section describes common **_<dfn>`u:` options</dfn>_** which each implementation SHOULD support
for all _functions_ and _markup_.

### `u:id`

A string value that is included as an `id` or other suitable value
in the formatted parts for the _placeholder_,
or any other structured formatted results.

Ignored when formatting a message to a string.

Accepts string values, or values which can be stringified without error.
For other values, a _Bad Option_ error is emitted
and the `u:id` option is ignored.

### `u:locale`

> A comma-delimited list of BCP 47 language tags,
> or an implementation-defined list of such tags.

Reviewer: The below text around line 33 suggests that that the language tags are supposed to be well-formed.

```suggestion
A comma-delimited list consisting of
well-formed [BCP 47](https://www.rfc-editor.org/rfc/bcp/bcp47.txt)
language tags,
or an implementation-defined list of such tags.
```

Replaces the _locale_ defined in the _function context_ for this _expression_.
The value is ignored when set on _markup_.

During processing, the `u:locale` option

> is always removed from the resolved mapping of _options_.

Reviewer: Tighten a bit?

```suggestion
is removed from the _resolved options_.
```

Author: Two thoughts here:
1. The "always" is intended to ensure that `u:locale` is removed from the options even if this bit from below applies:
   > For unsupported values, a _Bad Option_ error is emitted
   > and the value of the `u:locale` option is ignored.

   Without the "always", is it clear enough that it's always removed?
2. We don't define the term _resolved options_. This is what we have: https://github.com/unicode-org/message-format-wg/blob/e9fc1367cbf496897736e68b0a29fd3ace229759/spec/formatting.md?plain=1#L309-L310

Reviewer: WRT (1), I think the always is clear because "is removed" leaves no wiggle room. But I could live with always.

WRT (2), ah... I see. Perhaps:

```suggestion
MUST be removed from the _options_ in the _resolved value_.
```

That's stronger than _is_ and clarifies always. I do find that the phrasing is inelegant though and we could just ignore this comment.

Author: Applied some MUSTard.

Values matching the following ABNF are always accepted:
```abnf

> u-locale-option = langtag *([s] "," [s] langtag)

Reviewer: `[s]` is the same as `o`. Use that.

I don't agree with using `langtag` over using `Language-Tag`. Grandfathered tags all have modern replacements (or are deprecated outright) and private use is the user's problem. There's no reason for us to inject additional checking gorp onto the language tag space.

Author: ```suggestion
u-locale-option = langtag *(o "," o langtag)
```

Author: > There's no reason for us to inject additional checking gorp onto the language tag space.

I agree with this sentiment, but I think the current language is effectively required to allow for that. Note that we also include this further below:

> Implementations MAY support additional language tags,
> such as private-use or grandfathered tags,
> or tags using `_` instead of `-` as a separator.

With `langtag` as a base, we allow but do not require grandfathered or private-use tags to be supported, while also allowing but not requiring Unicode locale identifiers to be supported.

The overall intent here is to require the intersection of BCP47 and ULI to be supported, and allow extension from there, so that each implementation can use whatever is its native language tag format, and doesn't need to inject additional checking gorp onto its language tag space.

Reviewer: My tendency has been to require "well-formed" language tags (including allowing grandfathered and private-use) but encourage "valid" tags. Does the following capture the intention:

Implementations:
- MUST support well-formed language tags
- SHOULD check for valid language tags
- SHOULD canonicalize language tags and SHOULD ensure that tags are ULIs
- MAY omit, reject, or skip individual invalid tags
- MAY support additional implementation-defined identifiers
- MAY support implementation-defined locale objects
- MUST omit language tags from the resolved value

Users:
- SHOULD use valid canonicalized ULIs

I notice in making this list that we don't say whether the `language` attribute of the formatted value should be modified to match the selected `u:locale`.

Author: > MUST support well-formed language tags

This is problematic at least in a JS context, as it would require that e.g. a grandfathered tag like `i-klingon` is supported, and it currently is not: https://tc39.es/ecma402/#sec-language-tags

It is not reasonable to require an implementation to extend or modify its language tag rules to accommodate MF2.

Reviewer: You're right. Support is the wrong word, at least from the point of view of "locales that do something useful". My proposal should probably say "MUST accept well-formed language tags". (Note that canonicalization would turn `i-klingon` into `tlh`). The other SHOULD/MAY requirements allow well-formed but invalid tags to not be supported (consistent with EMCA-402). But a naive implementation could exist that accepted any seemingly-possible language tag (`xxx-Yyyy-ZZ-variant`)

Reviewer 2: I would stongly suggest referencing instead https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#unicode_bcp47_locale_id

It has a slightly narrower syntax than BCP47; see https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#bcp-47-conformance

(Note: once the spec goes final, these links will be modified to be of the form https://www.unicode.org/reports/tr35/xxx/tr35.html#yyy

Author: Referencing [`unicode_bcp47_locale_id`](https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#unicode_bcp47_locale_id) would be fine as well; it looks like it's effectively the same as the current `langtag`?

Reviewer: "MUST accept" means that any well-formed language tags can be passed in the language priority list. The range error in Intl happens later--it's allowed by "MAY omit, reject, or skip". The point is to be somewhat permissive about passing in things that might be language tags and then allowing implementations some leeway in disposing of them. In practice, no one is using `i-klingon` or `i-default` or `cel-gaulish`.

Author: Now I'm a bit confused. With what you propose, would it be acceptable for an implementation to emit a Bad Option error if `i-klingon` is used as a `u:locale` value?

Reviewer: Yes, although possibly not fatally.

The point of these attributes is to set the locale for some formatting. Ideally, we should allow any value that can appear in e.g. HTML's `lang` attribute. But we can't enforce that such a locale "works" or has data installed on a given implementation or system. So if you pass `u:locale=cel-gaulish` (or `u:locale=$celGaulish`), you probably get Bad Option and a formatted message using whatever the contextual locale is. FWIW, an implementation might do the same for non-range-erroring `aaa-Yyyy-RR` (not valid) or even the valid tag `tlh-Cyrl-AQ`. All of the canonicalization stuff we point to is in the service of making tags around the margins marginally useful.

Reviewer 2: What I propose is that 

1. There be a [ VC: ] constraint that the syntax for locale identifiers conforms to BCP47 so that we don't get complete garbage.
2. Implementations should disregard any locale identifier that they don't support (that is, not raise an error)
3. Implementations should follow the semantics of locale identifiers as in https://www.unicode.org/reports/tr35/tr35-general.html#unicode_bcp47_locale_id  
    - **NB: when moved to the CLDR spec, this should be a relative link tr35.html#unicode_bcp47_locale_id**

Reviewer: I can live with that.

Reviewer 2: @eemeli ?

Author: As I mentioned before, I'd be fine referencing `unicode_bcp47_locale_id` instead of `langtag`, but I'm not really sure how the further proposed changes would affect the text.

We already do not consider errors on options to be fatal, so I'd much rather not further restrict implementations from emitting errors for unsupported locales.

I'm going to apply an update, and then ask for another review pass.

```
using `langtag` as defined in [BCP 47](https://www.rfc-editor.org/rfc/bcp/bcp47.txt).
Note that `langtag` is the rule for "normal language tags",
and does not include private-use or grandfathered tags.

Implementations MAY support additional language tags,
such as private-use or grandfathered tags,
or tags using `_` instead of `-` as a separator.
When the value of `u:locale` is set by a _variable_,
implementations MAY support non-string values otherwise representing locales.

For unsupported values, a _Bad Option_ error is emitted
and the value of the `u:locale` option is ignored.

### `u:dir`

Replaces the base directionality defined in
the _function context_ for this _expression_.

> The value is ignored when set on _markup_.

Reviewer: This is probably a mistake. Instead of ignoring it, we should say the behavior is undefined (an implementation of markup might make use of it?)

Author: I disagree; see my previous reply to this: https://github.com/unicode-org/message-format-wg/pull/846#discussion_r1721885800.

Reviewer: In your previous comment, note that it doesn't prevent people from using it. This just says that it's not effective. My thinking is that we don't have a ton of implementation experience in this space. By not forbidding it (making it a Bad Option) we leave it in a kind of nether space ("valid" but ineffectual and tempting to abuse).

Author: I'd be fine emitting a Bad Option if this or `u:locale` is set on markup.

Author: Changed this to emit an error.

# MessageFormat 2.0 Unicode Namespace

> The `u:` namespace is reserved for use by the Unicode Consortium.

Reviewer: Here's where the definition goes, I guess? Perhaps:

```suggestion
The `u:` _namespace_ is reserved for the definition of _options_
for use in assisting the translation process 
or which affect the _function context_ of the specific _expressions_
in which they appear.
The CLDR Technical Committee of the Unicode Consortium
manages the specification for this namespace, hence the name `u:`.
```

Author: > for use in assisting the translation process

Isn't this handled by attributes?

Author: Leaving that out:
```suggestion
The `u:` _namespace_ is reserved for the definition of _options_
which affect the _function context_ of the specific _expressions_
in which they appear.

The CLDR Technical Committee of the Unicode Consortium
manages the specification for this namespace, hence the name `u:`.
```

This doesn't cover `u:id` on markup, though. Drawing a bit of a blank on exactly how to include that.

## Options

This section describes common **_<dfn>`u:` options</dfn>_** which each implementation SHOULD support
for all _functions_ and _markup_.

### `u:id`

A string value that is included as an `id` or other suitable value
in the formatted parts for the _placeholder_,
or any other structured formatted results.

Ignored when formatting a message to a string.

> Accepts string values, or values which can be stringified without error.

Reviewer: The option is not a function, so it doesn't really "accept" anything?

```suggestion
The value of the `id` _option_ can be a _literal_ or a
_variable_ whose _resolved value_ is either a string
or can be resolved to a string without error.
```

For other values, a _Bad Option_ error is emitted
and the `u:id` option is ignored.

### `u:locale`

A comma-delimited list consisting of
well-formed [BCP 47](https://www.rfc-editor.org/rfc/bcp/bcp47.txt)
language tags,
or an implementation-defined list of such tags.

> Replaces the _locale_ defined in the _function context_ for this _expression_.

Reviewer: The other two say "what" first and then tackle the contents. Move this to line 23?

> A comma-delimited list consisting of
> well-formed [BCP 47](https://www.rfc-editor.org/rfc/bcp/bcp47.txt)
> language tags,
> or an implementation-defined list of such tags.
> 
> Replaces the _locale_ defined in the _function context_ for this _expression_.

Author: ```suggestion
Replaces the _locale_ defined in the _function context_ for this _expression_.

A comma-delimited list consisting of
well-formed [BCP 47](https://www.rfc-editor.org/rfc/bcp/bcp47.txt)
language tags,
or an implementation-defined list of such tags.
```

If this option is set on _markup_, a _Bad Option_ error is emitted
and the value of the `u:locale` option is ignored.

During processing, the `u:locale` option
MUST be removed from the resolved mapping of _options_
before calling the _function handler_.

Values matching the following ABNF are always accepted:
```abnf
u-locale-option = unicode_bcp47_locale_id *(o "," o unicode_bcp47_locale_id)
```
using `unicode_bcp47_locale_id` as defined for
[Unicode Locale Identifier](https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#unicode_bcp47_locale_id).

Implementations MAY support additional language tags,
such as private-use or grandfathered tags,
or tags using `_` instead of `-` as a separator.
When the value of `u:locale` is set by a _variable_,
implementations MAY support non-string values otherwise representing locales.

> For unsupported values, a _Bad Option_ error is emitted
> and the value of the `u:locale` option is ignored.

Reviewer: Unsupported is unclear here. We also need to resolve list handling.

```
{$foo :date u:locale=|en-US, tlh-Xsux-AQ, i-klingon, bad-Juju, American|}`
```

Does the above work because en-US is fine? Or emit Bad Option because:
- `tlh-Xsux-AQ` a valid tag for Klingon as written in cuneiform as used in Antarctica is not an installed locale
- `i-klingon` is grandfathered and this implementation doesn't support grandfathered tags
- `bad-Juju` is not a valid tag
- `American` is not well-formed

```suggestion
If any of the language tags in the value is unsupported,
invalid, or not well-formed, the implementation ignores that entry in the
list and emits a _Bad Option_ error.
If the resulting list is empty, a _Bad Option_ error is emitted
and the value of the `u:locale` option is ignored.
```

Author: I fear that defining this too strictly mandate more than we'd like about how an implementation handles locales. Thinking about this some more, my preference would be to allow for each implementation to follow its own customs on locale fallback chain handling.

For instance JavaScript errors on not well-formed tags, but ignores invalid and unsupported ones:
```js
Intl.NumberFormat.supportedLocalesOf(['en-US', 'tlh-Xsux-AQ', 'i-klingon', 'bad-Juju'])
// Uncaught RangeError: invalid language tag: "i-klingon"

Intl.NumberFormat.supportedLocalesOf(['en-US', 'tok', 'bad-Juju', 'American'])
// [ "en-US" ]
```
but it's not reasonable to expect all other implementations to do the same thing.

Furthermore, referring to validity should probably be avoided as well, because in this context that would mean needing to carry around data that might not have any other use.

How about this?

```suggestion
Implementations MAY emit a _Bad Option_ error
and MAY ignore the value of the `u:locale` option
if it contains a locale that it does not support or
a language tag that is not well-formed.
```

Reviewer: Mt examples were meant as rhetorical. Your suggestion is fine, except that I think we should allow for the list to be processed as I suggested. I intended that as MAY... basically the implementation can process the list as it likes. But if the list has one bad tag in it, should we ignore the whole thing? That's how it reads now.

Author: The current language is intended to allow for an implementation to ignore the whole value and emit an error if it contains a bad tag, or to not do so, or to decide on a case-by-case basis. Note that we also do not specify how and when locale negotiation is meant to happen; that could be handled by the implementation, but it could also be done separately in each function handler.

Reviewer: (Back at a real computer)

Perhaps:

```suggestion
Implementations MAY emit a _Bad Option_ error
and MAY ignore the value of the `u:locale` _option_ as a whole
or any of the entries in the list of language tags.
This might be because the locale specified is not supported
or because the language tag is not well-formed,
not valid, or some other reason.
```

Author: Eh, seems a bit verbose but works.

# MessageFormat 2.0 Unicode Namespace

The `u:` _namespace_ is reserved for the definition of _options_
which affect the _function context_ of the specific _expressions_
in which they appear,
or for the definition of _options_ that are universally applicable

> rather than function-specific.

Reviewer 2: We also reserve `u:` functions (even if there are none currently).

Author: @aphillips Do you think we ought to call that out specifically?

Reviewer: Yeah, I think we should mention it. Perhaps add a sentence to the end like:

> It might also be used to define _functions_ in a future release.

Author: ```suggestion
rather than function-specific.
It might also be used to define _functions_ in a future release.
```
